File: sdk-python/obsvr/external_backend.py
"""Inbound external policy backend (ADR-4): OPA or Cedar.

Twin of sdk-typescript/src/policy/external-backend.ts. A customer points the SDK at their
existing policy-as-code engine (an OPA HTTP endpoint or a Cedar authorization
endpoint) and that engine's verdict participates in the pre-call decision.

Four guarantees (all pinned cross-language by
conformance/fixtures/external_backend.json):
  1. DENY-WINS merge with the local rules — a deny from EITHER side blocks; a
     backend "allow" never downgrades a local block.
  2. Fail-closed: a backend error OR timeout counts as DENY (enforce mode). A
     configurable shadow mode makes the backend observe-only (records what it
     WOULD have done, never blocks) for safe rollout.
  3. SSRF guard on the backend URL (see ssrf.py): non-http(s) schemes and
     private/loopback/link-local/metadata addresses are refused, resolving
     before connect. Any guard failure is an error outcome -> fail-closed.
  4. Provenance: the emitted event records which backend decided (identity + a
     hash of the effective backend policy) via the record returned here.

Zero-config default is NO backend (unchanged behavior). Stdlib only.
"""

import hashlib
import http.client
import json
import socket
import ssl
from typing import Any, Callable, Dict, List, Optional, Tuple
from urllib.parse import urlsplit

from .ssrf import (
    AllowedBackendAddress,
    AllowedBackendTarget,
    SsrfError,
    resolve_backend_url_allowed,
)

# Raw outcome before the shadow/fail-closed policy is applied. "error" and
# "timeout" are the fail-closed cases (deny in enforce mode); kept distinct from
# a genuine "deny" for provenance.
BackendOutcome = str  # "allow" | "deny" | "error" | "timeout"
LocalDecision = str  # "allow" | "redact" | "block"


def merge_external_backend_decision(
    local: LocalDecision, outcome: BackendOutcome, shadow: bool
) -> Dict[str, Any]:
    """DENY-WINS merge — the load-bearing, conformance-pinned function.

    A denial from EITHER side blocks; a backend "allow" never downgrades a local
    decision. ``error``/``timeout`` are denials in enforce mode (fail-closed). In
    shadow mode the backend NEVER changes the decision (observe-only), though the
    raw outcome is still recorded on the event.

    Returns {"decision", "blocked_by_backend"}.
    """
    backend_denies = outcome != "allow"  # deny | error | timeout
    if shadow:
        return {"decision": local, "blocked_by_backend": False}
    if backend_denies:
        return {"decision": "block", "blocked_by_backend": local != "block"}
    return {"decision": local, "blocked_by_backend": False}


def _url_host(url: str) -> str:
    """Host (with port when present), matching TS `new URL(url).host`."""
    try:
        parts = urlsplit(url)
    except ValueError:
        return url
    host = parts.hostname or url
    if parts.port is not None:
        host = f"{host}:{parts.port}"
    return host


def backend_provenance(cfg: Dict[str, Any]) -> Dict[str, str]:
    """Backend identity + effective-policy hash for provenance. The hash is a
    16-hex SHA-256 prefix of the configured policy text when present, else of the
    endpoint identity (``type|url``). Byte-identical to the TS twin."""
    btype = cfg["type"]
    url = cfg["url"]
    name = cfg.get("name")
    identity = name if isinstance(name, str) and name else f"{btype}:{_url_host(url)}"
    policy = cfg.get("policy")
    material = policy if isinstance(policy, str) and policy else f"{btype}|{url}"
    policy_hash = hashlib.sha256(material.encode("utf-8")).hexdigest()[:16]
    return {"identity": identity, "policy_hash": policy_hash}


def build_backend_input(
    *,
    operation: str,
    provider: str,
    model: str,
    environment: Optional[str] = None,
    user_id: Optional[str] = None,
    service_name: Optional[str] = None,
    tenant_id: Optional[str] = None,
    local_decision: LocalDecision,
    rules_hash: str,
    prompt_sha256: str,
) -> Dict[str, Any]:
    """Build the non-content decision-input document POSTed to the backend
    (digests, not raw prompts). Mirrors the TS buildBackendInput."""
    principal: Dict[str, str] = {}
    if user_id:
        principal["user_id"] = user_id
    if service_name:
        principal["service_name"] = service_name
    if tenant_id:
        principal["tenant_id"] = tenant_id
    doc: Dict[str, Any] = {
        "operation": operation,
        "provider": provider,
        "model": model,
    }
    if environment:
        doc["environment"] = environment
    doc["principal"] = principal
    doc["local_decision"] = local_decision
    doc["rules_hash"] = rules_hash
    doc["prompt_sha256"] = prompt_sha256
    return doc


def _normalize_opa(body: Any) -> Optional[Dict[str, Any]]:
    """Normalize an OPA ``result`` value into {"allow", "reasons"}."""
    if not isinstance(body, dict) or "result" not in body:
        return None  # an undefined document omits `result` -> caller: error
    result = body["result"]
    if isinstance(result, bool):
        return {"allow": result, "reasons": []}
    if isinstance(result, dict):
        reasons = [r for r in result.get("reasons", []) if isinstance(r, str)]
        if isinstance(result.get("allow"), bool):
            return {"allow": result["allow"], "reasons": reasons}
        if isinstance(result.get("deny"), bool):
            return {"allow": not result["deny"], "reasons": reasons}
        if isinstance(result.get("deny"), list):
            deny_reasons = [r for r in result["deny"] if isinstance(r, str)]
            return {
                "allow": len(result["deny"]) == 0,
                "reasons": deny_reasons or reasons,
            }
    return None


def _normalize_cedar(body: Any) -> Optional[Dict[str, Any]]:
    """Normalize a Cedar/AVP-style response into {"allow", "reasons"}."""
    if not isinstance(body, dict):
        return None
    decision = body.get("decision")
    if isinstance(decision, str):
        allow = decision.lower() == "allow"
        reasons: List[str] = []
        errors = body.get("errors")
        if errors is None and isinstance(body.get("diagnostics"), dict):
            errors = body["diagnostics"].get("errors")
        if isinstance(errors, list):
            for e in errors:
                if isinstance(e, str):
                    reasons.append(e)
                elif isinstance(e, dict) and isinstance(e.get("errorDescription"), str):
                    reasons.append(e["errorDescription"])
        return {"allow": allow, "reasons": reasons}
    return None


# Transport: fetch(url, headers, body, timeout_s) -> (status:int, parsed_json).
# Raises TimeoutError on timeout, any other Exception on error. This injectable
# seam is trusted to provide destination pinning equivalent to the production
# connector; the default retains the guard's approved address snapshot.
Transport = Callable[[str, Dict[str, str], str, float], Tuple[int, Any]]


def _connect_approved_socket(
    addresses: tuple[AllowedBackendAddress, ...], port: int, timeout_s: float
) -> socket.socket:
    """Connect directly to one approved numeric address without another lookup."""
    if timeout_s <= 0:
        raise TimeoutError("external backend timeout budget exhausted")
    last_error: Optional[OSError] = None
    for approved in addresses:
        sock = socket.socket(approved.family, socket.SOCK_STREAM)
        sock.settimeout(timeout_s)
        endpoint = (
            (approved.address, port)
            if approved.family == socket.AF_INET
            else (approved.address, port, 0, 0)
        )
        try:
            sock.connect(endpoint)
            return sock
        except OSError as exc:
            last_error = exc
            sock.close()
    if last_error is not None:
        raise last_error
    raise OSError("external backend has no approved address")


class _PinnedHTTPConnection(http.client.HTTPConnection):
    """One-shot HTTP connection whose socket destination is DNS-pinned."""

    def __init__(self, target: AllowedBackendTarget, timeout_s: float) -> None:
        host = target.parts.hostname or ""
        super().__init__(host, target.parts.port or 80, timeout=timeout_s)
        self._approved_addresses = target.addresses

    def connect(self) -> None:
        self.sock = _connect_approved_socket(
            self._approved_addresses, self.port, self.timeout
        )


class _PinnedHTTPSConnection(http.client.HTTPSConnection):
    """One-shot pinned HTTPS connection with origin-host TLS verification."""

    def __init__(
        self,
        target: AllowedBackendTarget,
        timeout_s: float,
        *,
        context: Optional[ssl.SSLContext] = None,
    ) -> None:
        host = target.parts.hostname or ""
        super().__init__(
            host,
            target.parts.port or 443,
            timeout=timeout_s,
            context=context,
        )
        self._approved_addresses = target.addresses

    def connect(self) -> None:
        raw_sock = _connect_approved_socket(
            self._approved_addresses, self.port, self.timeout
        )
        try:
            # The socket goes to an approved numeric address, but authentication
            # remains bound to the configured hostname (SNI + certificate check).
            self.sock = self._context.wrap_socket(raw_sock, server_hostname=self.host)
        except Exception:
            raw_sock.close()
            raise


def _host_header(target: AllowedBackendTarget) -> str:
    host = target.parts.hostname or ""
    if ":" in host:
        host = f"[{host}]"
    default_port = 443 if target.parts.scheme == "https" else 80
    if target.parts.port is not None and target.parts.port != default_port:
        return f"{host}:{target.parts.port}"
    return host


def _urllib_transport(
    target: AllowedBackendTarget,
    headers: Dict[str, str],
    body: str,
    timeout_s: float,
) -> Tuple[int, Any]:
    """POST through a fresh connection pinned to the validated DNS snapshot."""
    conn: http.client.HTTPConnection
    if target.parts.scheme == "https":
        conn = _PinnedHTTPSConnection(target, timeout_s)
    else:
        conn = _PinnedHTTPConnection(target, timeout_s)

    request_headers = {
        key: value
        for key, value in headers.items()
        if key.lower() not in {"host", "connection", "content-length"}
    }
    request_headers.update(
        {
            "Host": _host_header(target),
            "Connection": "close",
            "Content-Length": str(len(body.encode("utf-8"))),
        }
    )
    path = target.parts.path or "/"
    if target.parts.query:
        path = f"{path}?{target.parts.query}"
    try:
        conn.request("POST", path, body=body.encode("utf-8"), headers=request_headers)
        response = conn.getresponse()
        status = response.status
        raw = response.read()
    except (socket.timeout, TimeoutError) as exc:
        raise TimeoutError(str(exc)) from exc
    finally:
        # Never reuse a socket across validation snapshots.
        conn.close()
    try:
        return status, json.loads(raw)
    except (ValueError, TypeError):
        return status, None


def evaluate_external_backend(
    cfg: Dict[str, Any],
    input_doc: Dict[str, Any],
    *,
    transport: Optional[Transport] = None,
    resolver: Optional[Callable[[str], List[str]]] = None,
) -> Dict[str, Any]:
    """Consult the backend. NEVER raises — every failure mode (SSRF block,
    network error, non-2xx, unparseable body, timeout) maps to an
    ``error``/``timeout`` outcome so the caller's fail-closed merge stays in
    control. Returns {"outcome", "reasons"}."""
    allow_priv = bool(cfg.get("allow_private_network"))
    try:
        target = resolve_backend_url_allowed(cfg["url"], allow_priv, resolver)
    except SsrfError:
        return {"outcome": "error", "reasons": ["ssrf_guard_blocked_backend_url"]}
    except Exception:
        return {"outcome": "error", "reasons": ["ssrf_guard_blocked_backend_url"]}

    payload = {"input": input_doc} if cfg["type"] == "opa" else input_doc
    body = json.dumps(payload)
    headers = {"content-type": "application/json", **(cfg.get("headers") or {})}
    # `is None`, never `or`: an explicit timeout_ms of 0 is the operator
    # saying "no budget" (which fails closed as a timeout), and `or` would
    # silently replace that zero with a laxer 2-second default. The TS twin
    # uses `?? 2000` and honors 0 the same way.
    _timeout_ms = cfg.get("timeout_ms")
    timeout_s = (2000 if _timeout_ms is None else _timeout_ms) / 1000.0
    try:
        if transport is not None:
            status, parsed = transport(cfg["url"], headers, body, timeout_s)
        else:
            status, parsed = _urllib_transport(target, headers, body, timeout_s)
    except (socket.timeout, TimeoutError):
        return {"outcome": "timeout", "reasons": ["backend_timeout"]}
    except Exception:
        return {"outcome": "error", "reasons": ["backend_error"]}

    if not (200 <= int(status) < 300):
        return {"outcome": "error", "reasons": [f"backend_http_{status}"]}
    if parsed is None:
        return {"outcome": "error", "reasons": ["backend_response_not_json"]}

    normalized = _normalize_opa(parsed) if cfg["type"] == "opa" else _normalize_cedar(parsed)
    if normalized is None:
        return {"outcome": "error", "reasons": ["backend_response_unrecognized"]}
    return {
        "outcome": "allow" if normalized["allow"] else "deny",
        "reasons": normalized["reasons"],
    }


def run_external_backend_step(
    cfg: Dict[str, Any],
    local_decision: LocalDecision,
    input_doc: Dict[str, Any],
    *,
    transport: Optional[Transport] = None,
    resolver: Optional[Callable[[str], List[str]]] = None,
) -> Dict[str, Any]:
    """One-call integration step used by the pre-call pipeline: evaluate the
    backend, merge deny-wins, and assemble the provenance record. Called only
    when the local decision is not already a block. Returns
    {"decision", "blocked_by_backend", "record"}."""
    shadow = cfg.get("shadow") is True
    ev = evaluate_external_backend(cfg, input_doc, transport=transport, resolver=resolver)
    prov = backend_provenance(cfg)
    merge = merge_external_backend_decision(local_decision, ev["outcome"], shadow)
    record: Dict[str, Any] = {
        "identity": prov["identity"],
        "policy_hash": prov["policy_hash"],
        "type": cfg["type"],
        "outcome": ev["outcome"],
        "shadow": shadow,
    }
    if ev["reasons"]:
        record["reasons"] = ev["reasons"]
    return {
        "decision": merge["decision"],
        "blocked_by_backend": merge["blocked_by_backend"],
        "record": record,
    }

File: sdk-typescript/src/policy/external-backend.ts
/**
 * Inbound external policy backend (ADR-4): OPA or Cedar.
 *
 * Today Obsvr only EXPORTS to Rego one-way (rego-export.ts). This module adds
 * the INBOUND seam: a customer points the SDK at their existing policy-as-code
 * engine — an OPA HTTP endpoint or a Cedar authorization endpoint — and that
 * engine's verdict participates in the pre-call decision.
 *
 * Four guarantees (all pinned cross-language; twin: sdk-python/obsvr/external_backend.py):
 *   1. DENY-WINS merge with the local rules — a deny from EITHER side blocks.
 *      The backend can only ADD restriction; a backend "allow" never downgrades
 *      a local block. Pinned by conformance/fixtures/external_backend.json.
 *   2. Fail-closed: a backend error OR timeout counts as DENY (enforce mode),
 *      because a policy engine that cannot render a verdict must not be treated
 *      as approval. A configurable shadow mode makes the backend observe-only
 *      (records what it WOULD have done, never blocks) for safe rollout.
 *   3. SSRF guard on the backend URL (see utils/ssrf.ts): non-http(s) schemes and
 *      private/loopback/link-local/metadata addresses are refused, resolving
 *      every address once and pinning the socket to that approved snapshot. Any
 *      guard failure is an error outcome -> fail-closed. "Metadata addresses" means
 *      every IPv6 form that carries the v4 address, not only the mapped one:
 *      until this was measured, `http://[::169.254.169.254]/` reached `fetch`
 *      while `::ffff:169.254.169.254` was refused.
 *   4. Provenance: the emitted event records which backend decided (identity +
 *      a hash of the effective backend policy) via the ExternalBackendRecord.
 *
 * Zero-config default is NO backend (unchanged behavior).
 *
 * @packageDocumentation
 */

import { createHash } from 'node:crypto';
import { createPinnedLookup, postPinnedBytes } from '../utils/pinned-http.js';
import {
  resolveBackendUrlAllowed,
  type AllowedBackendTarget,
  type Resolver,
  type SsrfOptions,
} from '../utils/ssrf.js';

/** Which policy-as-code engine the backend speaks. */
export type ExternalBackendType = 'opa' | 'cedar';

/**
 * Raw outcome of consulting the backend, before the shadow/fail-closed policy
 * is applied. `error` and `timeout` are the fail-closed cases (treated as deny
 * in enforce mode); they are kept distinct from a genuine `deny` for provenance.
 */
export type BackendOutcome = 'allow' | 'deny' | 'error' | 'timeout';

/** Local decision reached by the rest of the pre-call pipeline. */
export type LocalDecision = 'allow' | 'redact' | 'block';

/** Customer configuration for the inbound external policy backend. */
export interface ExternalPolicyBackendConfig {
  /** 'opa' (POST {input}, read result.allow/bool) or 'cedar' (read decision Allow/Deny). */
  type: ExternalBackendType;
  /**
   * Full decision endpoint URL. For OPA this is the data document, e.g.
   * `https://opa.example.com/v1/data/obsvr/allow`. Must be http(s); SSRF-guarded.
   */
  url: string;
  /** Observe-only: record the verdict but never block. Default false (enforce). */
  shadow?: boolean;
  /** Per-call budget (ms). Error/timeout => deny in enforce mode. Default 2000. */
  timeoutMs?: number;
  /** Extra request headers (auth tokens, etc.). */
  headers?: Record<string, string>;
  /** Human label recorded on events (provenance). Default `${type}:${host}`. */
  name?: string;
  /**
   * The effective backend policy text/identity (e.g. the Rego module, the Cedar
   * policy set, or a bundle revision id). Hashed into the event's policy hash so
   * the audit record ties to a specific backend policy version. When omitted,
   * the hash is derived from the endpoint identity instead.
   */
  policy?: string;
  /** Permit a loopback/private-network backend (never metadata/link-local). Default false. */
  allowPrivateNetwork?: boolean;
}

/** Non-content decision input POSTed to the backend (digests, not raw prompts). */
export interface BackendDecisionInput {
  operation: string;
  provider: string;
  model: string;
  environment?: string;
  /** Principal/subject identifiers visible at the boundary (only non-empty ones). */
  principal: { user_id?: string; service_name?: string; tenant_id?: string };
  /** The local pipeline's decision before the backend was consulted. */
  local_decision: LocalDecision;
  /** Canonical rules hash (policy_version) the local decision ran under. */
  rules_hash: string;
  /** SHA-256 of the evaluated prompt text (digest — the raw prompt is never sent). */
  prompt_sha256: string;
}

/**
 * Provenance record stamped on the emitted audit event: WHICH backend decided,
 * and what it said. Additive (never part of the HMAC chain preimage).
 */
export interface ExternalBackendRecord {
  /** Human identity of the backend (config.name or `${type}:${host}`). */
  identity: string;
  /** 16-hex SHA-256 prefix of the effective backend policy (or endpoint identity). */
  policy_hash: string;
  type: ExternalBackendType;
  /** Raw outcome before shadow/fail-closed. */
  outcome: BackendOutcome;
  /** True when observe-only (the outcome did not affect the decision). */
  shadow: boolean;
  /** Reasons returned by the backend, when any. */
  reasons?: string[];
}

/** Result of the deny-wins merge between the local decision and the backend. */
export interface BackendMergeResult {
  decision: LocalDecision;
  /** True iff the backend's denial is what produced the block. */
  blocked_by_backend: boolean;
}

/**
 * DENY-WINS merge — the load-bearing, conformance-pinned function.
 *
 * A denial from EITHER side blocks; a backend "allow" never downgrades a local
 * decision. `error`/`timeout` are denials in enforce mode (fail-closed). In
 * shadow mode the backend NEVER changes the decision (observe-only), though the
 * raw outcome is still recorded on the event for rollout visibility.
 */
export function mergeExternalBackendDecision(
  local: LocalDecision,
  outcome: BackendOutcome,
  shadow: boolean,
): BackendMergeResult {
  const backendDenies = outcome !== 'allow'; // deny | error | timeout
  if (shadow) {
    return { decision: local, blocked_by_backend: false };
  }
  if (backendDenies) {
    return { decision: 'block', blocked_by_backend: local !== 'block' };
  }
  return { decision: local, blocked_by_backend: false };
}

/** Host portion of a URL for identity, or the raw url on parse failure. */
function urlHost(url: string): string {
  try {
    return new URL(url).host;
  } catch {
    return url;
  }
}

/**
 * Backend identity + effective-policy hash for provenance. The hash is a
 * 16-hex SHA-256 prefix of the configured policy text when present, else of the
 * endpoint identity (`type|url`). Deterministic and cross-language-identical.
 */
export function backendProvenance(cfg: ExternalPolicyBackendConfig): {
  identity: string;
  policy_hash: string;
} {
  const identity = cfg.name && cfg.name.length > 0 ? cfg.name : `${cfg.type}:${urlHost(cfg.url)}`;
  const material =
    cfg.policy && cfg.policy.length > 0 ? cfg.policy : `${cfg.type}|${cfg.url}`;
  const policy_hash = createHash('sha256').update(material, 'utf8').digest('hex').slice(0, 16);
  return { identity, policy_hash };
}

/** Build the non-content decision-input document POSTed to the backend. */
export function buildBackendInput(params: {
  operation: string;
  provider: string;
  model: string;
  environment?: string;
  userId?: string;
  serviceName?: string;
  tenantId?: string;
  localDecision: LocalDecision;
  rulesHash: string;
  promptSha256: string;
}): BackendDecisionInput {
  const principal: BackendDecisionInput['principal'] = {};
  if (params.userId) principal.user_id = params.userId;
  if (params.serviceName) principal.service_name = params.serviceName;
  if (params.tenantId) principal.tenant_id = params.tenantId;
  return {
    operation: params.operation,
    provider: params.provider,
    model: params.model,
    ...(params.environment ? { environment: params.environment } : {}),
    principal,
    local_decision: params.localDecision,
    rules_hash: params.rulesHash,
    prompt_sha256: params.promptSha256,
  };
}

/** Normalize an OPA `result` value into allow + reasons. */
function normalizeOpa(body: unknown): { allow: boolean; reasons: string[] } | null {
  if (!body || typeof body !== 'object') return null;
  if (!('result' in body)) return null; // OPA with an undefined document omits `result`
  const result = (body as { result: unknown }).result;
  if (typeof result === 'boolean') return { allow: result, reasons: [] };
  if (result && typeof result === 'object') {
    const r = result as Record<string, unknown>;
    const reasons = Array.isArray(r.reasons)
      ? (r.reasons.filter((x) => typeof x === 'string') as string[])
      : [];
    if (typeof r.allow === 'boolean') return { allow: r.allow, reasons };
    if (typeof r.deny === 'boolean') return { allow: !r.deny, reasons };
    if (Array.isArray(r.deny)) {
      const denyReasons = r.deny.filter((x) => typeof x === 'string') as string[];
      return { allow: r.deny.length === 0, reasons: denyReasons.length ? denyReasons : reasons };
    }
  }
  return null; // unrecognized shape -> caller treats as error (fail-closed)
}

/** Normalize a Cedar/AVP-style response into allow + reasons. */
function normalizeCedar(body: unknown): { allow: boolean; reasons: string[] } | null {
  if (!body || typeof body !== 'object') return null;
  const b = body as Record<string, unknown>;
  const decision = b.decision;
  if (typeof decision === 'string') {
    const allow = decision.toLowerCase() === 'allow';
    const reasons: string[] = [];
    const errors = b.errors ?? (b.diagnostics as Record<string, unknown> | undefined)?.errors;
    if (Array.isArray(errors)) {
      for (const e of errors) {
        if (typeof e === 'string') reasons.push(e);
        else if (e && typeof e === 'object' && typeof (e as { errorDescription?: unknown }).errorDescription === 'string') {
          reasons.push((e as { errorDescription: string }).errorDescription);
        }
      }
    }
    return { allow, reasons };
  }
  return null;
}

/** Dependencies injectable for tests. */
export interface EvaluateDeps {
  /**
   * Explicit transport seam for tests/embedders. The production path does not
   * use fetch: it pins the socket to the address snapshot returned by the SSRF
   * guard. A supplied fetch implementation is trusted to provide equivalent
   * destination pinning and must not be used with an untrusted URL.
   */
  fetchImpl?: typeof fetch;
  resolver?: Resolver;
}

interface BackendHttpResponse {
  ok: boolean;
  status: number;
  json(): Promise<unknown>;
  type?: string;
}

export const createPinnedBackendLookup = createPinnedLookup;

/**
 * POST through Node's native HTTP stack with a pinned lookup. Passing the
 * original URL preserves its Host header. For HTTPS, `servername` is kept as
 * the configured hostname, so certificate validation remains hostname-based
 * even though the socket connects to the approved numeric address.
 */
async function postPinnedJson(
  target: AllowedBackendTarget,
  body: string,
  headers: Record<string, string>,
  signal: AbortSignal,
): Promise<BackendHttpResponse> {
  const response = await postPinnedBytes(target, body, headers, signal, 1_048_576);
  return {
    ok: response.status >= 200 && response.status < 300,
    status: response.status,
    json: async () => JSON.parse(Buffer.from(response.body ?? []).toString('utf8')),
  };
}

/**
 * Consult the backend. NEVER throws — every failure mode (SSRF block, network
 * error, non-2xx, unparseable body, timeout) maps to an `error`/`timeout`
 * outcome so the caller's fail-closed merge stays in control.
 */
export async function evaluateExternalBackend(
  cfg: ExternalPolicyBackendConfig,
  input: BackendDecisionInput,
  deps: EvaluateDeps = {},
): Promise<{ outcome: BackendOutcome; reasons: string[] }> {
  const ssrfOpts: SsrfOptions = { allowPrivateNetwork: cfg.allowPrivateNetwork };
  let target: AllowedBackendTarget;
  try {
    target = await resolveBackendUrlAllowed(cfg.url, ssrfOpts, deps.resolver);
  } catch {
    return { outcome: 'error', reasons: ['ssrf_guard_blocked_backend_url'] };
  }

  const timeoutMs = cfg.timeoutMs ?? 2000;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  if (typeof (timer as unknown as { unref?: () => void }).unref === 'function') {
    (timer as unknown as { unref: () => void }).unref();
  }

  // OPA reads the decision under `input`; Cedar receives the document directly.
  const payload = cfg.type === 'opa' ? { input } : input;
  const requestBody = JSON.stringify(payload);
  const requestHeaders = { 'content-type': 'application/json', ...(cfg.headers ?? {}) };

  try {
    const resp: BackendHttpResponse = deps.fetchImpl
      ? await deps.fetchImpl(cfg.url, {
          method: 'POST',
          headers: requestHeaders,
          body: requestBody,
          signal: controller.signal,
          redirect: 'manual',
        })
      : await postPinnedJson(target, requestBody, requestHeaders, controller.signal);
    if (!resp.ok || (resp as { type?: string }).type === 'opaqueredirect') {
      return { outcome: 'error', reasons: [`backend_http_${resp.status}`] };
    }
    let body: unknown;
    try {
      body = await resp.json();
    } catch {
      return { outcome: 'error', reasons: ['backend_response_not_json'] };
    }
    const normalized = cfg.type === 'opa' ? normalizeOpa(body) : normalizeCedar(body);
    if (!normalized) {
      return { outcome: 'error', reasons: ['backend_response_unrecognized'] };
    }
    return { outcome: normalized.allow ? 'allow' : 'deny', reasons: normalized.reasons };
  } catch (err) {
    const isAbort = err instanceof Error && (err.name === 'AbortError' || err.name === 'TimeoutError');
    return { outcome: isAbort ? 'timeout' : 'error', reasons: [isAbort ? 'backend_timeout' : 'backend_error'] };
  } finally {
    clearTimeout(timer);
  }
}

/** Outcome of the single external-backend step: merged decision + provenance. */
export interface ExternalBackendStepResult {
  decision: LocalDecision;
  blocked_by_backend: boolean;
  record: ExternalBackendRecord;
}

/**
 * One-call integration step used by the wrapper and integration pre-call paths:
 * evaluate the backend, merge deny-wins, and assemble the provenance record.
 * The caller invokes this only when the local decision is not already a block
 * (a block cannot be downgraded, so consulting the backend would be pure
 * overhead — and the deny-wins outcome is already decided).
 */
export async function runExternalBackendStep(
  cfg: ExternalPolicyBackendConfig,
  localDecision: LocalDecision,
  input: BackendDecisionInput,
  deps: EvaluateDeps = {},
): Promise<ExternalBackendStepResult> {
  const shadow = cfg.shadow === true;
  const { outcome, reasons } = await evaluateExternalBackend(cfg, input, deps);
  const prov = backendProvenance(cfg);
  const merge = mergeExternalBackendDecision(localDecision, outcome, shadow);
  const record: ExternalBackendRecord = {
    identity: prov.identity,
    policy_hash: prov.policy_hash,
    type: cfg.type,
    outcome,
    shadow,
    ...(reasons.length ? { reasons } : {}),
  };
  return { decision: merge.decision, blocked_by_backend: merge.blocked_by_backend, record };
}
